src/flb_encoding.c

```diff
@@ -0,0 +1,138 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019      The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <stdio.h>
+
+#include <string.h>
+#include <time.h>
+#include <ctype.h>
+
+#include <fluent-bit/flb_macros.h>
+#include <fluent-bit/flb_config.h>
+#include <fluent-bit/flb_error.h>
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/flb_str.h>
+#include <fluent-bit/flb_encoding.h>
+
+
+#include <tutf8e.h>
+
+/*
+ *
+ *  flb_encoding_open(encoding):
+ *  iso-8859-1,...
+ *  windows-1251 windows-1252, ..
+ * 
+ *  <charset>                     - fail if bad chars
+ *  <charset>//IGNORE             - ignore bad chars
```

Comment:
is the ```//``` just a separator or a common naming across encoding configs ?

Reply:
iconv(3)   library uses same convention to add  parameters to  charset.
implementations I know understand only   //IGNORE  and //TRANSLIT .

Reply:
I'm a bit sceptical of following the iconv API for this.
I think a nullable _invalid_ UTF8 string is sufficient for configuration purposes.
I don't think fluent-bit should be concerned with //TRANSLIT mode in particular:

```
ICONV_OPEN(3)                                               Linux Programmer's Manual                                               
...
NAME
       iconv_open - allocate descriptor for character set conversion

...
       //TRANSLIT
              When the string "//TRANSLIT" is appended to tocode, transliteration is activated.  This means that when a character cannot be  rep‐
              resented in the target character set, it can be approximated through one or several similarly looking characters.

       //IGNORE
              When  the  string  "//IGNORE"  is  appended  to  tocode,  characters that cannot be represented in the target character set will be
              silently discarded.```

Reply:
//TRANSLIT has no meaning in our case. 
//IGNORE is usefull...

Reply:
So ... we are going to get some kind agreement of configuration?

I like idea of one configuration value with //(parameter), because that is flexible and  "kind of"  compatible with iconv.
Also it's useful to have multiple ways to handle bad input (ignore,fail,replace-with-something).

Reply:
I don't think the iconv formulation is particularly modern, concise or appropriate for `fluent-bit` (in my opinion).

Reply:
So...  what this "modern" could mean?

Reply:
So how this should be configured???????

Reply:
I propose that the plugin supports an optional UTF8 _invalid_ string, along the lines of tutf8e.
//IGNORE would simply be an empty string, default might be ? or similar.

Reply:
Well...  that is this code doing... just putting everything to one parameter (like but not totally same as in iconv(3)). 
But should this be an another parameter????

Then probably better that replacement is should be (json) escaped string.
`

Encoding  iso-8859-2               => default... FAIL,INGORE ,  "?", "\uFFFD"  ????
Encoding  iso-8859-2  ?            => 
Encoding  iso-8859-3  <bad\20char> => "<bad char
Encoding  iso-8859-3  \uFFFD       => Use unicode replacement character

Maybe  

Encoding  iso-8859-2  \I            =>  IGNORE
Encoding  iso-8859-2  \F            =>  FAIL
Encoding  iso-8859-2  \R            =>  Replacement char...
`
Or we put optional second parameter to quotes "?" ...
hum... getting complicated.

Reply:
Are getting anywhere with configuration????  @nigels-com

Reply:
I'll take a fresh look.  I know you came to this from an iconv perspective, I was hoping to leave that behind rather than carry that into fluent-bit.  It's been a while since I was actively engaged on this one.

Reply:
hum.. maybe have to change way of configuration...

Reply:
Changed configuration  use have parameters: 

* **encoding**  CHARSET
* **encoding_replacement**  STRING  

default is to use unicode replacement char (0xfffd)